### ISSUE TITLE
Lint Idea: Call may cause panic

### ISSUE THREAD

**Issue by jamesmunns** [rust-lang-nursery/rust-clippy#2942]
It would be nice to have a lint that would inform me if a call I make could panic, and ideally if possible, recommend an alternative that uses `Option` or `Result` to handle the error directly. For example, with the following input:

```rust
fn main() {
    use std::time::Duration;
    
    let x = Duration::from_secs(2);
    let y = Duration::from_secs(1);
    
    println!("{:?}", y - x);
}
```

I would like a warning that looks something like this:

```
src/main.rs
   |
 7 | println!("{:?}", y - x);
   |                    ^ warning: Sub<Duration> for Duration may panic due to this:

src/core/time.rs:428
     | fn sub(self, rhs: Duration) -> Duration {
 428 |     self.checked_sub(rhs).expect("overflow when subtracting durations")
     | }
     |                          ^^^^^^^ call to `expect` could fail

Consider using `Duration::checked_sub()` instead
```

I imagine this could fall into "Halting problem" territory, but I imagine the process would go something like:

* find every operator or function call for a given program (you know, no big deal)
* obtain the full possible branching graph for that call/operator (also nbd)
* look for `unwrap`/`expect`/`panic`
* Check if any steps in the call graph to get to the panic have some kind of recommended alternative. Maybe possible to cover the std lib, though supporting external libraries/versions would be difficult to say the least. If no alternative, at least warn

I'm very open to hear things like "this is possible, but difficult", "this is impossible, here is where we discussed previously", etc.

**Comment by Manishearth** [comment]
We basically don't do global analyses in clippy, aside from being hard (generics are tricky) and expensive, they're impossible with the current rustc lint design because we run on individual crates and don't have this kind of metadata from dependencies.

I believe https://github.com/llogiq/metacollect has the goal of doing stuff  like this, though the project seems abandoned.

**Comment by jamesmunns** [comment]
Thanks for the feedback @Manishearth, I submitted it since I didn't see any previous issues that exactly discussed this, though https://github.com/rust-lang-nursery/rust-clippy/issues/959 and https://github.com/rust-lang-nursery/rust-clippy/issues/2536 are somewhat related. I somewhat expected it to get rejected :)

Also I'll check out metacollect if I get some idle cycles.

**Comment by flip1995** [comment]
You may also be interested in https://github.com/Technolution/rustig. This tool checks if your Rust program has any paths leading to the panic handler.
